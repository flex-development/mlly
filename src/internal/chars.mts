/**
 * @file Internal - chars
 * @module mlly/internal/chars
 */

/**
 * Characters that may have special meaning.
 *
 * @enum {string}
 */
enum chars {
  asterisk = '*',
  at = '@',
  empty = '',
  hash = '#'
}

export default chars
