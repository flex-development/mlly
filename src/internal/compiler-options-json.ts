/**
 * @file Internals - CompilerOptionsJson
 * @module mlly/internal/CompilerOptionsJson
 */

import type { CompilerOptionsValue, MapLike } from 'typescript'

/**
 * TypeScript compiler options provided by a user.
 *
 * @internal
 *
 * @see https://www.typescriptlang.org/tsconfig
 */
interface CompilerOptionsJson {
  allowJs?: boolean
  allowSyntheticDefaultImports?: boolean
  allowUmdGlobalAccess?: boolean
  allowUnreachableCode?: boolean
  allowUnusedLabels?: boolean
  alwaysStrict?: boolean
  assumeChangesOnlyAffectDirectDependencies?: boolean
  baseUrl?: string
  charset?: string
  checkJs?: boolean
  composite?: boolean
  declaration?: boolean
  declarationDir?: string
  declarationMap?: boolean
  disableReferencedProjectLoad?: boolean
  disableSizeLimit?: boolean
  disableSolutionSearching?: boolean
  disableSourceOfProjectReferenceRedirect?: boolean
  downlevelIteration?: boolean
  emitBOM?: boolean
  emitDeclarationOnly?: boolean
  emitDecoratorMetadata?: boolean
  esModuleInterop?: boolean
  exactOptionalPropertyTypes?: boolean
  experimentalDecorators?: boolean
  forceConsistentCasingInFileNames?: boolean
  importHelpers?: boolean
  importsNotUsedAsValues?: 'error' | 'preserve' | 'remove'
  incremental?: boolean
  inlineSourceMap?: boolean
  inlineSources?: boolean
  isolatedModules?: boolean
  jsx?: 'preserve' | 'react-jsx' | 'react-jsxdev' | 'react-native' | 'react'
  jsxFactory?: string
  jsxFragmentFactory?: string
  jsxImportSource?: string
  keyofStringsOnly?: boolean
  lib?: (
    | 'DOM.Iterable'
    | 'dom.iterable'
    | 'DOM'
    | 'dom'
    | 'ES5'
    | 'es5'
    | 'ES6'
    | 'es6'
    | 'ES7'
    | 'es7'
    | 'ES2015.Collection'
    | 'es2015.collection'
    | 'ES2015.Core'
    | 'es2015.core'
    | 'ES2015.Generator'
    | 'es2015.generator'
    | 'ES2015.Iterable'
    | 'es2015.iterable'
    | 'ES2015.Promise'
    | 'es2015.promise'
    | 'ES2015.Proxy'
    | 'es2015.proxy'
    | 'ES2015.Reflect'
    | 'es2015.reflect'
    | 'ES2015.Symbol.WellKnown'
    | 'es2015.symbol.wellknown'
    | 'ES2015.Symbol'
    | 'es2015.symbol'
    | 'ES2015'
    | 'es2015'
    | 'ES2016.Array.Include'
    | 'es2016.array.include'
    | 'ES2016'
    | 'es2016'
    | 'ES2017.Intl'
    | 'es2017.intl'
    | 'ES2017.Object'
    | 'es2017.object'
    | 'ES2017.SharedMemory'
    | 'es2017.sharedmemory'
    | 'ES2017.String'
    | 'es2017.string'
    | 'ES2017.TypedArrays'
    | 'es2017.typedarrays'
    | 'ES2017'
    | 'es2017'
    | 'ES2018.AsyncGenerator'
    | 'es2018.asyncgenerator'
    | 'ES2018.AsyncIterable'
    | 'es2018.asynciterable'
    | 'ES2018.Intl'
    | 'es2018.intl'
    | 'ES2018.Promise'
    | 'es2018.promise'
    | 'ES2018.Regexp'
    | 'es2018.regexp'
    | 'ES2018'
    | 'es2018'
    | 'ES2019.Array'
    | 'es2019.array'
    | 'ES2019.Object'
    | 'es2019.object'
    | 'ES2019.String'
    | 'es2019.string'
    | 'ES2019.Symbol'
    | 'es2019.symbol'
    | 'ES2019'
    | 'es2019'
    | 'ES2020.BigInt'
    | 'es2020.bigint'
    | 'ES2020.Intl'
    | 'es2020.intl'
    | 'ES2020.Promise'
    | 'es2020.promise'
    | 'ES2020.SharedMemory'
    | 'es2020.sharedmemory'
    | 'ES2020.String'
    | 'es2020.string'
    | 'ES2020.Symbol.WellKnown'
    | 'es2020.symbol.wellknown'
    | 'ES2020'
    | 'es2020'
    | 'ES2021.Promise'
    | 'es2021.promise'
    | 'ES2021.String'
    | 'es2021.string'
    | 'ES2021.WeakRef'
    | 'es2021.weakref'
    | 'ES2021'
    | 'es2021'
    | 'ESNext.Array'
    | 'esnext.array'
    | 'ESNext.AsyncIterable'
    | 'esnext.asynciterable'
    | 'ESNext.BigInt'
    | 'esnext.bigint'
    | 'ESNext.Intl'
    | 'esnext.intl'
    | 'ESNext.Promise'
    | 'esnext.promise'
    | 'ESNext.String'
    | 'esnext.string'
    | 'ESNext.Symbol'
    | 'esnext.symbol'
    | 'ESNext.WeakRef'
    | 'esnext.weakref'
    | 'ESNext'
    | 'esnext'
    | 'ScriptHost'
    | 'scripthost'
    | 'WebWorker.ImportScripts'
    | 'webworker.importscripts'
    | 'WebWorker.Iterable'
    | 'webworker.iterable'
    | 'WebWorker'
    | 'webworker'
  )[]
  locale?: string
  mapRoot?: string
  maxNodeModuleJsDepth?: number
  module?:
    | 'AMD'
    | 'amd'
    | 'CommonJS'
    | 'commonjs'
    | 'ES6'
    | 'es6'
    | 'ES2015'
    | 'es2015'
    | 'ES2020'
    | 'es2020'
    | 'ES2022'
    | 'es2022'
    | 'ESNext'
    | 'esnext'
    | 'Node16'
    | 'node16'
    | 'NodeNext'
    | 'nodenext'
    | 'None'
    | 'none'
    | 'System'
    | 'system'
    | 'UMD'
    | 'umd'
  moduleDetection?: 'auto' | 'force' | 'legacy'
  moduleResolution?: 'classic' | 'node' | 'node16' | 'nodenext'
  moduleSuffixes?: string[]
  newLine?: 'CRLF' | 'crlf' | 'LF' | 'lf'
  noEmit?: boolean
  noEmitHelpers?: boolean
  noEmitOnError?: boolean
  noErrorTruncation?: boolean
  noFallthroughCasesInSwitch?: boolean
  noImplicitAny?: boolean
  noImplicitOverride?: boolean
  noImplicitReturns?: boolean
  noImplicitThis?: boolean
  noImplicitUseStrict?: boolean
  noLib?: boolean
  noPropertyAccessFromIndexSignature?: boolean
  noResolve?: boolean
  noStrictGenericChecks?: boolean
  noUncheckedIndexedAccess?: boolean
  noUnusedLocals?: boolean
  noUnusedParameters?: boolean
  out?: string
  outDir?: string
  outFile?: string
  paths?: MapLike<string[]>
  preserveConstEnums?: boolean
  preserveSymlinks?: boolean
  preserveValueImports?: boolean
  project?: string
  reactNamespace?: string
  removeComments?: boolean
  resolveJsonModule?: boolean
  rootDir?: string
  rootDirs?: string[]
  skipDefaultLibCheck?: boolean
  skipLibCheck?: boolean
  sourceMap?: boolean
  sourceRoot?: string
  strict?: boolean
  strictBindCallApply?: boolean
  strictFunctionTypes?: boolean
  strictNullChecks?: boolean
  strictPropertyInitialization?: boolean
  stripInternal?: boolean
  suppressExcessPropertyErrors?: boolean
  suppressImplicitAnyIndexErrors?: boolean
  target?:
    | 'ES3'
    | 'es3'
    | 'ES5'
    | 'es5'
    | 'ES6'
    | 'es6'
    | 'ES2015'
    | 'es2015'
    | 'ES2016'
    | 'es2016'
    | 'ES2017'
    | 'es2017'
    | 'ES2018'
    | 'es2018'
    | 'ES2019'
    | 'es2019'
    | 'ES2020'
    | 'es2020'
    | 'ES2021'
    | 'es2021'
    | 'ESNext'
    | 'esnext'
  traceResolution?: boolean
  tsBuildInfoFile?: string
  typeRoots?: string[]
  types?: string[]
  useDefineForClassFields?: boolean
  useUnknownInCatchVariables?: boolean
  [option: string]: CompilerOptionsValue | undefined
}

export type { CompilerOptionsJson as default }
