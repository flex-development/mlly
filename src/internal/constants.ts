/**
 * @file Internals - Constants
 * @module mlly/internal/constants
 */

/**
 * CommonJS syntax regex.
 *
 * @internal
 *
 * @const {RegExp} CJS_SYNTAX_REGEX
 */
export const CJS_SYNTAX_REGEX: RegExp =
  /(?<!(?:\/\/|\*).*)((?:module\.)?exports(?:\.\w+|(?<!\s+=))|require(?=\(.)|require\.\w+|__dirname|__filename|await import)/gm

/**
 * Dynamic import statement regex.
 *
 * @internal
 *
 * @const {RegExp} DYNAMIC_IMPORT_REGEX
 */
export const DYNAMIC_IMPORT_REGEX: RegExp =
  /(?<!(?:\/\/|\*).*)(?:(?:const|let|var)(?:(?<name>\s+\w+\s*)|\s*{\s*(?<names>\w+(?:,\s*\w+)*)\s*}\s*)(?:\s*=\s*?)?.)?(?:await)?\s+import\s*\((?<expression>(?:[^()]+|\((?:[^()]+|\([^()]*\))*\))*)\)/gm

/**
 * ESM syntax regex.
 *
 * @internal
 *
 * @const {RegExp} ESM_SYNTAX_REGEX
 */
export const ESM_SYNTAX_REGEX: RegExp =
  /(?<!(?:\/\/|\*).*)((?:export|import)[\s\w*,{}]*(?=\sfrom)|export\b\s*(?:[*{]|async function|(?:abstract\s)?class|const|default|enum|function|interface|let|type|var)|await import|import\.meta\.(?:env(?:\.\w+)?|resolve|url))/gm

/**
 * Declaration export statement regex.
 *
 * **Note**: Captures [declaration type][1] and export name only.
 *
 * [1]: ./types/declaration.ts
 *
 * @const {RegExp} EXPORT_DECLARATION_REGEX
 */
export const EXPORT_DECLARATION_REGEX: RegExp =
  /(?<!(?:\/\/|\*).*)\bexport(?: +declare)? +(?<declaration>(?:(?:abstract +)?class|(?:default +)?async +function\*?|(?:default +)?function\*?|default +async|default(?!(?: +async)? +function\*?)|(?:const +)?enum|const|interface|let|namespace|type(?! *{)|var)) +(?<name>[\w$]+(?!.*\b +=>))?/gm

/**
 * Named export statement regex.
 *
 * @internal
 *
 * @const {RegExp} EXPORT_NAMED_REGEX
 */
export const EXPORT_NAMED_REGEX: RegExp =
  /(?<!(?:\/\/|\*).*)\bexport(?:\s+type)?\s+{(?<exports>[^}]+?)[\s,]*}(?:\s*from\s*["']\s*(?<specifier>(?<="\s*)[^"]*[^\s"](?=\s*")|(?<='\s*)[^']*[^\s'](?=\s*'))\s*["'][^\n;]*)?/gm

/**
 * Aggregate export statement (`export * ...`) regex.
 *
 * @internal
 *
 * @const {RegExp} EXPORT_STAR_REGEX
 */
export const EXPORT_STAR_REGEX: RegExp =
  /(?<!(?:\/\/|\*).*)\bexport\s*\*(?:\s*as\s+(?<name>[\w$]+)\s+)?\s*(?:\s*from\s*["']\s*(?<specifier>(?<="\s*)[^"]*[^\s"](?=\s*")|(?<='\s*)[^']*[^\s'](?=\s*'))\s*["'][^\n;]*)?/gm

/**
 * Require statement regex.
 *
 * @internal
 *
 * @const {RegExp} REQUIRE_STATEMENT_REGEX
 */
export const REQUIRE_STATEMENT_REGEX: RegExp =
  /(?<!(?:\/\/|\*).*)(?:\bconst[ {]+(?<imports>[\w\t\n\r $*,/]+)[ =}]+)?(?<type>\brequire(?:\.resolve)?)\(["'] *(?<specifier>(?<="\s*)[^"']*[^\s"'](?=\s*")|(?<=["']\s*)[^']*[^\s'](?=\s*')) *["']\)/gm

/**
 * Static import statement regex.
 *
 * @internal
 *
 * @const {RegExp} STATIC_IMPORT_REGEX
 */
export const STATIC_IMPORT_REGEX: RegExp =
  /(?<!(?:\/\/|\*).*)import\s*(?:[\s"']*(?<imports>[\w\t\n\r $*,/{}]+)from\s*)?["']\s*(?<specifier>(?<="\s*)[^"]*[^\s":](?=\s*")|(?<=["']\s*)[^']*[^\s':](?=\s*'))\s*["']\s*/gm
