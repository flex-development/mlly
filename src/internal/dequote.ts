/**
 * @file Internal - dequote
 * @module mlly/internal/dequote
 */

import { trim } from '@flex-development/tutils'

/**
 * Removes quotes from the beginning and end of a string.
 *
 * @internal
 *
 * @param {string} str - String to dequote
 * @return {string} String with quotes removed
 */
const dequote = (str: string): string => trim(str).replace(/^["']|["']$/g, '')

export default dequote
