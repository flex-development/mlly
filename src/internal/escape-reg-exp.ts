/**
 * @file Internal - escapeRegExp
 * @module mlly/internal/escapeRegExp
 */

import type { NodeError } from '@flex-development/errnode'
import validateString from './validate-string'

/**
 * Escapes special characters in the given regex `pattern`.
 *
 * A backslash escape (`\\`) is used when valid. A `\x2d` escape is used when
 * the former would be disallowed by stricter Unicode pattern grammar.
 *
 * @param {string} pattern - Regex pattern to escape
 * @return {string} `pattern` with special characters escaped
 * @throws {NodeError<TypeError>} If `pattern` is not a string
 */
const escapeRegExp = (pattern: string): string => {
  validateString(pattern, 'pattern')
  return pattern.replace(/[$()*+.?[\\\]^{|}]/g, '\\$&').replace(/-/g, '\\x2d')
}

export default escapeRegExp
