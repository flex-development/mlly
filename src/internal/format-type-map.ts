/**
 * @file Internal - FORMAT_TYPE_MAP
 * @module mlly/internal/FORMAT_TYPE_MAP
 */

import { AssertType, Format } from '#src/enums'

/**
 * Module format to `import` assertion type map.
 *
 * **Note**:  It's unclear whether or not the HTML spec will require assertion
 * types for WebAssembly modules.
 *
 * @see https://github.com/WebAssembly/esm-integration/issues/42
 *
 * @const {Map<Format, AssertType>} FORMAT_TYPE_MAP
 */
const FORMAT_TYPE_MAP: Map<Format, AssertType> = new Map<Format, AssertType>([
  [Format.BUILTIN, AssertType.IMPLICIT],
  [Format.COMMONJS, AssertType.IMPLICIT],
  [Format.JSON, AssertType.JSON],
  [Format.MODULE, AssertType.IMPLICIT],
  [Format.WASM, AssertType.IMPLICIT]
])

export default FORMAT_TYPE_MAP
