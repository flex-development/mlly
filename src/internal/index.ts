/**
 * @file Internals
 * @module mlly/internal
 */

export { default as getCompilerOptions } from './get-compiler-options'
