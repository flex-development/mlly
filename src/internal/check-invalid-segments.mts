/**
 * @file Internal - checkInvalidSegments
 * @module mlly/internal/checkInvalidSegments
 */

import chars from '#internal/chars'

/**
 * Check `str` for invalid path segments.
 *
 * @internal
 *
 * @param {string} str
 *  The package target or subpath pattern match to check
 * @return {boolean}
 *  `true` if `str` contains invalid path segments, `false` otherwise
 */
function checkInvalidSegments(str: string): boolean {
  /**
   * Regular expression matching path separators.
   *
   * @const {RegExp} separator
   */
  const separator: RegExp = /\/|\\/

  /**
   * Path segments.
   *
   * @const {string[]} segments
   */
  const segments: string[] = str.startsWith('./')
    ? str.slice(2).split(separator)
    : str.split(separator)

  /**
   * Regular expressions matching invalid segments.
   *
   * @const {RegExp[]} checks
   */
  const checks: RegExp[] = [
    /^(\.|%2e)(\.|%2e)?/i,
    /^(n|%6e|%4e)(o|%6f|%4f)(d|%64|%44)(e|%65|%45)(_|%5f)(m|%6d|%4d)(o|%6f|%4f)(d|%64|%44)(u|%75|%55)(l|%6c|%4c)(e|%65|%45)(s|%73|%53)/i
  ]

  for (const segment of segments) {
    if (segment === chars.empty || checks.some(check => check.test(segment))) {
      return true
    }
  }

  return false
}

export default checkInvalidSegments
