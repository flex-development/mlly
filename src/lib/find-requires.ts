/**
 * @file findRequires
 * @module mlly/lib/findRequires
 */

import type { RequireStatement } from '#src/interfaces'
import { REQUIRE_STATEMENT_REGEX } from '#src/internal'

/**
 * Finds all `require` and `require.resolve` statements in `code`.
 *
 * @param {string} code - Code to check
 * @return {RequireStatement[]} Require statement objects
 */
const findRequires = (code: string): RequireStatement[] => {
  return [...code.matchAll(REQUIRE_STATEMENT_REGEX)].map(match => {
    const { 0: statement = '', index: start = 0, groups = {} } = match
    const { imports = '', specifier = '', type = '' } = groups

    return {
      code: statement,
      end: start + statement.length,
      imports:
        imports === '' || type === 'require.resolve'
          ? []
          : /const *\w/.test(statement)
          ? ['default']
          : imports.split(',').map(name => name.trim()),
      specifier,
      start,
      type: type as RequireStatement['type']
    }
  })
}

export default findRequires
