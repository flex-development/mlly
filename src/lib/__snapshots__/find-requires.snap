// Vitest Snapshot v1

exports[`unit:lib/findRequires > should return RequireStatement object array 1`] = `
[
  {
    "code": "require('./side-effect.cjs')",
    "end": 28,
    "imports": [],
    "kind": "require",
    "specifier": "./side-effect.cjs",
    "specifier_kind": "static",
    "start": 0,
    "syntax": "require",
  },
  {
    "code": "const mod=require('./lib')",
    "end": 55,
    "imports": [
      "mod",
    ],
    "kind": "require",
    "specifier": "./lib",
    "specifier_kind": "static",
    "start": 29,
    "syntax": "require",
  },
  {
    "code": "const{createPathMatch}=require('tsconfig-paths')",
    "end": 105,
    "imports": [
      "createPathMatch",
    ],
    "kind": "require",
    "specifier": "tsconfig-paths",
    "specifier_kind": "static",
    "start": 57,
    "syntax": "require",
  },
  {
    "code": "const mod = require('lib')",
    "end": 133,
    "imports": [
      "mod",
    ],
    "kind": "require",
    "specifier": "lib",
    "specifier_kind": "static",
    "start": 107,
    "syntax": "require",
  },
  {
    "code": "require(id)",
    "end": 146,
    "imports": [],
    "kind": "require",
    "specifier": "id",
    "specifier_kind": "dynamic",
    "start": 135,
    "syntax": "require",
  },
  {
    "code": "require('#src/lib/resolve-aliases')",
    "end": 182,
    "imports": [],
    "kind": "require",
    "specifier": "#src/lib/resolve-aliases",
    "specifier_kind": "static",
    "start": 147,
    "syntax": "require",
  },
  {
    "code": "const {
  default: myDefault,
  addFive,
	addFour,
	addThree,
	addTwo,
	squareFive,
	squareFour,
	squareThree,
	squareTwo
} = require('./lib')",
    "end": 325,
    "imports": [
      "default: myDefault",
      "addFive",
      "addFour",
      "addThree",
      "addTwo",
      "squareFive",
      "squareFour",
      "squareThree",
      "squareTwo",
    ],
    "kind": "require",
    "specifier": "./lib",
    "specifier_kind": "static",
    "start": 183,
    "syntax": "require",
  },
]
`;
